# Penilaian Kredit

# Daftar Isi <a id='back'></a>

* [Pendahuluan](#intro)
* [Tahap 1. Ikhtisar Data](#data_review)
    * [Kesimpulan](#data_review_conclusions)
* [Tahap 2. Pra-pemrosesan data](#data_preprocessing)
    * [2.1 Mengelola kolom 'children'](#kol_children)
    * [2.2 Mengelola kolom 'days_employed' & 'total_income'](#kol_daysemp_totinc)
    * [2.3 Mengelola kolom 'dob_years'](#kol_dobyears)
    * [2.4 Mengelola kolom 'education' & 'education_id'](#kol_education)
    * [2.5 Mengelola kolom 'family_status_id' & 'family_status'](#kol_famstat)
    * [2.6 Mengelola kolom 'gender'](#kol_gender)
    * [2.7 Mengelola kolom 'income_type' - 'debt' - 'purpose'](#kol_inctype_debt_purpose)
    * [2.8 Menangani duplikasi data](#duplikasi_data)
    * [2.9 Mengategorikan data](#kategori_data)
    * [2.10 Kesimpulan](#data_preprocessing_conclusions)
* [Tahap 3. Analisis Hubungan](#analisis_hubungan)
    * [3.1 Hubungan 1: memiliki anak dan probabilitas gagal bayar](#analisis_hub_1)
    * [3.2 Hubungan 2: status perkawinan dan probabilitas gagal bayar](#analisis_hub_2)
    * [3.3 Hubungan 3: tingkat pendapatan dan probabilitas gagal bayar](#analisis_hub_3)
    * [3.4 Hubungan 4: tujuan pinjaman dan probabilitas gagal bayar](#analisis_hub_4)
* [Temuan](#end)

## Pendahuluan <a id='intro'></a>

Proyek ini ditujukan untuk menyiapkan laporan untuk bank bagian kredit. Dari laporan akan dicari hubungan beberapa faktor  terhadap probabilitas gagal bayar dalam melunasi pinjaman. Pihak bank sudah memiliki beberapa data mengenai kelayakan kredit nasabah.
Laporan Anda akan menjadi bahan pertimbangan pada saat membuat **penilaian** kredit untuk calon nasabah. **Penilaian kredit** digunakan untuk mengevaluasi kemampuan calon peminjam untuk melunasi pinjaman mereka.

### Tujuan: 
Menguji empat hubungan:
1. Apakah terdapat hubungan antara memiliki anak dan probabilitas seseorang melakukan gagal bayar pinjaman?
2. Apakah terdapat hubungan antara status perkawinan dan probabilitas seseorang melakukan gagal bayar pinjaman?
3. Apakah terdapat hubungan antara tingkat pendapatan dan probabilitas seseorang melakukan gagal bayar pinjaman?
4. Bagaimana perbedaan tujuan pinjaman memengaruhi probabilitas seseorang melakukan gagal bayar pinjaman?


### Tahapan
Data tentang nasabah pinjaman disimpan dalam *file* `/datasets/credit_scoring_eng.csv`. Tidak ada informasi terkait kualitas data tersebut, jadi perlu diperiksa terlebih dahulu sebelum menguji hipotesis.

Pertama, Anda akan mengevaluasi kualitas data dan melihat apakah masalahnya signifikan. Kemudian, selama pra-pemrosesan data, Anda akan mencoba mengatasi masalah yang paling serius.
 
Proyek ini akan terdiri dari tiga tahap:
 1. Ikhtisar Data
 2. Pra-pemrosesan Data
 3. Analisis Hubungan

 
[Kembali ke Daftar Isi](#back)

## Tahap 1. Ikhtisar Data <a id='data_review'></a>

Buka data terkait nasabah pinjaman, lalu jelajahilah data tersebut.

**Step 1.1**
Melakukan import library yang diperlukan

In [1]:
import pandas as pd

**Step 1.2**
Melakukan load data yang diperlukan dari file *credit_scoring_eng.csv*

In [2]:
try:
    df = pd.read_csv('credit_scoring_eng.csv')
except:
    df = pd.read_csv('datasets/credit_scoring_eng.csv')

**Step 1.3**
Memeriksa informasi umum dari dataframe yang di-*load* file *credit_scoring_eng.csv*

In [3]:
df.head()

children  days_employed  dob_years            education  education_id  \
0         1   -8437.673028         42    bachelor's degree             0   
1         1   -4024.803754         36  secondary education             1   
2         0   -5623.422610         33  Secondary Education             1   
3         3   -4124.747207         32  secondary education             1   
4         0  340266.072047         53  secondary education             1   

       family_status  family_status_id gender income_type  debt  total_income  \
0            married                 0      F    employee     0     40620.102   
1            married                 0      F    employee     0     17932.802   
2            married                 0      M    employee     0     23341.752   
3            married                 0      M    employee     0     42820.568   
4  civil partnership                 1      F     retiree     0     25378.572   

                   purpose  
0    purchase of the house  
1             car purchase  
2    purchase of the house  
3  supplementary education  
4        to have a wedding

In [4]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 21525 entries, 0 to 21524
Data columns (total 12 columns):
 #   Column            Non-Null Count  Dtype  
---  ------            --------------  -----  
 0   children          21525 non-null  int64  
 1   days_employed     19351 non-null  float64
 2   dob_years         21525 non-null  int64  
 3   education         21525 non-null  object 
 4   education_id      21525 non-null  int64  
 5   family_status     21525 non-null  object 
 6   family_status_id  21525 non-null  int64  
 7   gender            21525 non-null  object 
 8   income_type       21525 non-null  object 
 9   debt              21525 non-null  int64  
 10  total_income      19351 non-null  float64
 11  purpose           21525 non-null  object 
dtypes: float64(2), int64(5), object(5)
memory usage: 2.0+ MB


In [5]:
df.describe()

children  days_employed     dob_years  education_id  \
count  21525.000000   19351.000000  21525.000000  21525.000000   
mean       0.538908   63046.497661     43.293380      0.817236   
std        1.381587  140827.311974     12.574584      0.548138   
min       -1.000000  -18388.949901      0.000000      0.000000   
25%        0.000000   -2747.423625     33.000000      1.000000   
50%        0.000000   -1203.369529     42.000000      1.000000   
75%        1.000000    -291.095954     53.000000      1.000000   
max       20.000000  401755.400475     75.000000      4.000000   

       family_status_id          debt   total_income  
count      21525.000000  21525.000000   19351.000000  
mean           0.972544      0.080883   26787.568355  
std            1.420324      0.272661   16475.450632  
min            0.000000      0.000000    3306.762000  
25%            0.000000      0.000000   16488.504500  
50%            0.000000      0.000000   23202.870000  
75%            1.000000      0.000000   32549.611000  
max            4.000000      1.000000  362496.645000

### Kesimpulan <a id='data_review_conclusions'></a> 

Setiap baris dalam tabel menyimpan data terkait kredit yang telah diberikan kepada nasabah. Sejumlah kolom menyimpan data yang mendeskripsikan nasabah, yakni: jumlah anak, lama waktu bekerja, tingkat pendidikan, status pernikahan, gender, pekerjaan, dan pendapatan. Sisanya menyimpan data terkait informasi pinjaman, yakni: tujuan pinjaman dan status pinjaman gagal bayar.

Terlihat bahwa data yang dimiliki cukup untuk menganalisis hubungan. Meski demikian, kita memiliki nilai-nilai yang hilang, nilai-nilai negatif yang tidak pada tempatnya, dan nilai-nilai yang tidak rasional.

Untuk melanjutkan analisis, kita perlu melakukan pra-pemrosesan data terlebih dahulu.

[Kembali ke Daftar Isi](#back)

## Tahap 2. Pra-pemrosesan Data <a id='data_preprocessing'></a>
Memperbaiki nilai yang hilang, nilai negatif yang tidak pada tempatnya, dan nilai yang tidak rasional. Selain itu juga memperbaiki duplikat baik implisit maupun eksplisit.

### Mengelola kolom 'children' <a id='kol_children'></a>

**Step 2.1.1**

Mengetahui variasi dan jumlah data dari kolom 'children' yang diurutkan berdasarkan index.

In [6]:
df['children'].value_counts().sort_index()

-1        47
 0     14149
 1      4818
 2      2055
 3       330
 4        41
 5         9
 20       76
Name: children, dtype: int64

**Step 2.1.2**

Dari proses sebelumnya ditemukan bahwa terdapat 2 item klasifikasi data yang anomali yakni **-1** dan **20**.

Proses dilanjutkan dengan melihat secara umum data yang dimaksud.

In [7]:
df_test = df.loc[df['children'] == -1]
df_test

children  days_employed  dob_years            education  education_id  \
291          -1   -4417.703588         46  secondary education             1   
705          -1    -902.084528         50  secondary education             1   
742          -1   -3174.456205         57  secondary education             1   
800          -1  349987.852217         54  secondary education             1   
941          -1            NaN         57  Secondary Education             1   
1363         -1   -1195.264956         55  SECONDARY EDUCATION             1   
1929         -1   -1461.303336         38  secondary education             1   
2073         -1   -2539.761232         42  secondary education             1   
3814         -1   -3045.290443         26  Secondary Education             1   
4201         -1    -901.101738         41  secondary education             1   
4402         -1  398001.302888         64  SECONDARY EDUCATION             1   
4542         -1   -1811.899756         32  secondary education             1   
5272         -1   -1143.485347         46  secondary education             1   
6013         -1   -1361.258696         46    bachelor's degree             0   
6385         -1  370215.476226         48  secondary education             1   
7190         -1   -5928.202068         34  SECONDARY EDUCATION             1   
7278         -1    -526.318451         51  secondary education             1   
7615         -1            NaN         35  secondary education             1   
7685         -1   -3237.360455         53  secondary education             1   
8229         -1   -1803.441074         54    bachelor's degree             0   
8251         -1   -3113.998449         31  secondary education             1   
8561         -1    -195.479496         31  secondary education             1   
9565         -1   -2896.629686         34    bachelor's degree             0   
9574         -1   -2710.419901         28    bachelor's degree             0   
10177        -1   -1743.604011         37  secondary education             1   
10365        -1  345774.125957         63    BACHELOR'S DEGREE             0   
10918        -1  340499.039342         50  secondary education             1   
11109        -1   -3438.463024         59  secondary education             1   
11268        -1   -1048.782203         30    bachelor's degree             0   
11617        -1    -370.827130         27  secondary education             1   
12304        -1   -8493.101252         61         some college             2   
13786        -1            NaN         42  secondary education             1   
14357        -1    -268.337037         23  secondary education             1   
15147        -1   -9851.184337         44  secondary education             1   
15435        -1   -3614.220232         40    bachelor's degree             0   
16129        -1    -457.861760         33  secondary education             1   
16265        -1   -2802.218127         40    bachelor's degree             0   
17095        -1   -2809.693200         34  secondary education             1   
17429        -1    -895.379738         37  secondary education             1   
17657        -1   -4571.957475         41  secondary education             1   
18219        -1   -3575.215641         33  secondary education             1   
19100        -1    -617.246968         38  secondary education             1   
19366        -1    -372.034749         43  SECONDARY EDUCATION             1   
19417        -1  350340.760224         28  secondary education             1   
20393        -1  355157.107212         69  secondary education             1   
20717        -1    -661.822321         32    bachelor's degree             0   
21140        -1   -1422.668059         44  secondary education             1   

           family_status  family_status_id gender    income_type  debt  \
291    civil partnership                 1      F       employee     0   
705              m

In [8]:
df_test = df.loc[df['children'] == 20]
df_test

children  days_employed  dob_years            education  education_id  \
606          20    -880.221113         21  secondary education             1   
720          20    -855.595512         44  secondary education             1   
1074         20   -3310.411598         56  secondary education             1   
2510         20   -2714.161249         59    bachelor's degree             0   
2941         20   -2161.591519          0  secondary education             1   
...         ...            ...        ...                  ...           ...   
21008        20   -1240.257910         40  secondary education             1   
21325        20    -601.174883         37  secondary education             1   
21390        20            NaN         53  secondary education             1   
21404        20    -494.788448         52  secondary education             1   
21491        20    -173.954460         27  secondary education             1   

         family_status  family_status_id gender income_type  debt  \
606            married                 0      M    business     0   
720            married                 0      F    business     0   
1074           married                 0      F    employee     1   
2510   widow / widower                 2      F    employee     0   
2941           married                 0      F    employee     0   
...                ...               ...    ...         ...   ...   
21008          married                 0      F    employee     1   
21325          married                 0      F    business     0   
21390          married                 0      M    business     0   
21404          married                 0      M    business     0   
21491          married                 0      F    employee     0   

       total_income                                   purpose  
606       23253.578                     purchase of the house  
720       18079.798                           buy real estate  
1074      36722.966                      getting an education  
2510      42315.974  transactions with commercial real estate  
2941      31958.391                              to buy a car  
...             ...                                       ...  
21008     21363.842                              to own a car  
21325     16477.771                         profile education  
21390           NaN               buy residential real estate  
21404     25060.749          transactions with my real estate  
21491     24883.344                  buying a second-hand car  

[76 rows x 12 columns]

**Step 2.1.3**

Dari proses tersebut di atas secara sekilas kita juga dapat melihat permasalahan data di kolom yang lain yang nantinya akan secara bertahap ditindaklanjuti, antara lain nilai minus pada kolom days_employed, inkonsistensi label klasifikasi pada kolom education, dan juga beberapa hal yang terkait dengan *outlier* yang tidak rasional.

In [9]:
df['children'].describe()

count    21525.000000
mean         0.538908
std          1.381587
min         -1.000000
25%          0.000000
50%          0.000000
75%          1.000000
max         20.000000
Name: children, dtype: float64

**Step 2.1.4**

Melakukan penggantian nilai anomali dengan nilai yang rasional :
* nilai -1 diganti dengan nilai absolutnya yakni 1 dengan pertimbangan kemungkinan terdapat typo penginputan yang disertai dengan karakter '-'
* nilai 20 diganti dengan nilai 2 dengan pertimbangan jumlah anak pada data children sebelum lompat ke 20 adalah 5 sehingga kemungkinan besar angka 20 dimaksudkan untuk berada antara 0-5 serta juga kemungkinan terdapat typo penginputan yang disertai dengan karakter '0'

In [10]:
df['children'] = df['children'].replace({-1:1, 20:2})

In [11]:
df['children'].value_counts().sort_index()

0    14149
1     4865
2     2131
3      330
4       41
5        9
Name: children, dtype: int64

In [12]:
df['children'].describe()

count    21525.000000
mean         0.479721
std          0.755528
min          0.000000
25%          0.000000
50%          0.000000
75%          1.000000
max          5.000000
Name: children, dtype: float64

### Mengelola kolom 'days_employed' & 'total_income' <a id='kol_daysemp_totinc'></a>

**Step 2.2.1**

Mengetahui variasi dan jumlah data dari kolom 'days_employed' & 'total_income' yang diurutkan berdasarkan index, serta mendapatkan informasi statistik deskriptif dari kolom-kolom tersebut.

In [13]:
df['days_employed'].value_counts().sort_index()

-18388.949901     1
-17615.563266     1
-16593.472817     1
-16264.699501     1
-16119.687737     1
                 ..
 401663.850046    1
 401674.466633    1
 401675.093434    1
 401715.811749    1
 401755.400475    1
Name: days_employed, Length: 19351, dtype: int64

In [14]:
df['total_income'].value_counts().sort_index()

3306.762      1
3392.845      1
3418.824      1
3471.216      1
3503.298      1
             ..
273809.483    1
274402.943    1
276204.162    1
352136.354    1
362496.645    1
Name: total_income, Length: 19348, dtype: int64

In [15]:
df['days_employed'].describe()

count     19351.000000
mean      63046.497661
std      140827.311974
min      -18388.949901
25%       -2747.423625
50%       -1203.369529
75%        -291.095954
max      401755.400475
Name: days_employed, dtype: float64

In [16]:
df['total_income'].describe()

count     19351.000000
mean      26787.568355
std       16475.450632
min        3306.762000
25%       16488.504500
50%       23202.870000
75%       32549.611000
max      362496.645000
Name: total_income, dtype: float64

Dari penelusuran di atas ditemukan bahwa :
* terdapat nilai minus pada kolom 'days_employed' dan hal ini perlu diganti dengan nilai absolutnya
* terdapat outlier baik pada kolom 'days_employed' maupun 'total_income'
* outlier pada 'days_employed' nilainya tidak rasional dan harus diganti dengan nilai yang lain

**Step 2.2.2**

Menindaklanjuti nilai minus dengan mengubahnya menjadi nilai absolut

In [17]:
df['days_employed'] = abs(df['days_employed'])

In [18]:
df['days_employed'].value_counts().sort_index()

24.141633        1
24.240695        1
30.195337        1
33.520665        1
34.701045        1
                ..
401663.850046    1
401674.466633    1
401675.093434    1
401715.811749    1
401755.400475    1
Name: days_employed, Length: 19351, dtype: int64

**Step 2.2.3**

Mendeteksi nilai yang hilang

In [19]:
print(df.isna().sum())

children               0
days_employed       2174
dob_years              0
education              0
education_id           0
family_status          0
family_status_id       0
gender                 0
income_type            0
debt                   0
total_income        2174
purpose                0
dtype: int64


In [20]:
df_nan = df.loc[df['days_employed'].isna()]
df_nan

children  days_employed  dob_years            education  education_id  \
12            0            NaN         65  secondary education             1   
26            0            NaN         41  secondary education             1   
29            0            NaN         63  secondary education             1   
41            0            NaN         50  secondary education             1   
55            0            NaN         54  secondary education             1   
...         ...            ...        ...                  ...           ...   
21489         2            NaN         47  Secondary Education             1   
21495         1            NaN         50  secondary education             1   
21497         0            NaN         48    BACHELOR'S DEGREE             0   
21502         1            NaN         42  secondary education             1   
21510         2            NaN         28  secondary education             1   

           family_status  family_status_id gender    income_type  debt  \
12     civil partnership                 1      M        retiree     0   
26               married                 0      M  civil servant     0   
29             unmarried                 4      F        retiree     0   
41               married                 0      F  civil servant     0   
55     civil partnership                 1      F        retiree     1   
...                  ...               ...    ...            ...   ...   
21489            married                 0      M       business     0   
21495  civil partnership                 1      F       employee     0   
21497            married                 0      F       business     0   
21502            married                 0      F       employee     0   
21510            married                 0      F       employee     0   

       total_income                   purpose  
12              NaN         to have a wedding  
26              NaN                 education  
29              NaN    building a real estate  
41              NaN  second-hand car purchase  
55              NaN         to have a wedding  
...             ...                       ...  
21489           NaN         purchase of a car  
21495           NaN          wedding ceremony  
21497           NaN       building a property  
21502           NaN    building a real estate  
21510           NaN              car purchase  

[2174 rows x 12 columns]

**Step 2.2.4**

Mengganti nilai yang hilang

In [21]:
print(df[['days_employed','total_income']].agg('mean'))
print(df[['days_employed','total_income']].agg('median'))
print(df[['days_employed','total_income']].agg('std'))

days_employed    66914.728907
total_income     26787.568355
dtype: float64
days_employed     2194.220567
total_income     23202.870000
dtype: float64
days_employed    139030.880527
total_income      16475.450632
dtype: float64


Mengingat terdapat outlier yang mempengaruhi signifikan nilai rata-rata, maka yang dipilih untuk menggantikan nilai yang hilang (juga nantinya nilai outlier) adalah median.

In [22]:
df['days_employed'] = df['days_employed'].fillna(df['days_employed'].median())
df['total_income'] = df['total_income'].fillna(df['total_income'].median())
print(df.isna().sum())

children            0
days_employed       0
dob_years           0
education           0
education_id        0
family_status       0
family_status_id    0
gender              0
income_type         0
debt                0
total_income        0
purpose             0
dtype: int64


In [23]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 21525 entries, 0 to 21524
Data columns (total 12 columns):
 #   Column            Non-Null Count  Dtype  
---  ------            --------------  -----  
 0   children          21525 non-null  int64  
 1   days_employed     21525 non-null  float64
 2   dob_years         21525 non-null  int64  
 3   education         21525 non-null  object 
 4   education_id      21525 non-null  int64  
 5   family_status     21525 non-null  object 
 6   family_status_id  21525 non-null  int64  
 7   gender            21525 non-null  object 
 8   income_type       21525 non-null  object 
 9   debt              21525 non-null  int64  
 10  total_income      21525 non-null  float64
 11  purpose           21525 non-null  object 
dtypes: float64(2), int64(5), object(5)
memory usage: 2.0+ MB


**Step 2.2.5**

Mengubah tipe data 'float' menjadi 'int'

In [24]:
df = df.astype({'days_employed':'int', 'total_income':'int'})
df.dtypes

children             int64
days_employed        int32
dob_years            int64
education           object
education_id         int64
family_status       object
family_status_id     int64
gender              object
income_type         object
debt                 int64
total_income         int32
purpose             object
dtype: object

**Step 2.2.6**

Manganalisis dan menindaklanjuti outlier

In [25]:
df['days_employed'].value_counts()

2194      2180
133         16
327         16
438         15
223         14
          ... 
8200         1
9090         1
360849       1
2101         1
343937       1
Name: days_employed, Length: 9086, dtype: int64

In [26]:
df['days_employed'].value_counts().sort_index()

24        2
30        1
33        1
34        1
37        1
         ..
401663    1
401674    1
401675    1
401715    1
401755    1
Name: days_employed, Length: 9086, dtype: int64

**Notes:**

Merasionalisasi data kolom 'days_employed' dengan data pada kolom 'dob_years'.

Data angka 15.000 adalah angka yang menunjukkan bekerja selama kurang lebih 42 tahun, bila dirasionalisasikan dengan kolom 'dob_years' atau usia, maka usia tidak boleh melebihi 'days_employed' atau waktu bekerja.

In [27]:
df_test = df.loc[df['days_employed'] > 15000].sort_index()
print(df_test[['days_employed','dob_years']].sort_values('days_employed').head(30))

       days_employed  dob_years
5581           15079         55
16868          15193         57
19788          15267         56
15675          15410         65
7731           15618         64
4321           15773         61
1539           15785         59
3974           15835         64
16825          16119         64
17838          16264         59
7329           16593         60
4299           17615         61
16335          18388         61
20444         328728         72
9328          328734         41
17782         328771         56
14783         328795         62
7229          328827         32
5332          328834         55
5756          328842         51
13325         328862         55
3642          328874         54
3386          328895         58
15491         328909         60
7153          328911         44
14247         328923         57
13929         328928         54
7711          328959         62
8516          328994         61
2575          329001         51


**Notes:**

Angka terbesar yang masih rasional adalah 18.388 atau masa kerja selama kurang lebih 49 tahun. Data di bawah angka tersebut dengan bantuan perbandingan kolom 'dob_years' masih dalam kewajaran. Selebihnya outlier yang tidak rasional pada kolom 'days_employed' yakni setelah angka 18.388. Oleh karenanya, angka di atas 18.388 perlu diganti dengan angka median.

In [28]:
print(df.loc[df['days_employed'] > 18388, 'days_employed'].count())
df.loc[df['days_employed'] > 18388, 'days_employed'] = df['days_employed'].median()

3445


In [29]:
print(df.loc[df['days_employed'] > 18388].sort_index())


Empty DataFrame
Columns: [children, days_employed, dob_years, education, education_id, family_status, family_status_id, gender, income_type, debt, total_income, purpose]
Index: []


**Step 2.2.7**

Memeriksa kembali kolom 'total_income'.

In [30]:
df['total_income'].value_counts()

23202    2175
19552       7
23344       6
16686       5
24821       5
         ... 
41514       1
45228       1
34624       1
27232       1
35966       1
Name: total_income, Length: 15388, dtype: int64

In [31]:
df['total_income'].value_counts().sort_index()

3306      1
3392      1
3418      1
3471      1
3503      1
         ..
273809    1
274402    1
276204    1
352136    1
362496    1
Name: total_income, Length: 15388, dtype: int64

### Mengelola kolom 'dob_years' <a id='kol_dobyears'></a>

**Step 2.3.1**

Mengetahui variasi dan jumlah data dari kolom 'dob_years' yang diurutkan berdasarkan index, serta mendapatkan informasi statistik deskriptif dari kolom-kolom tersebut.

In [32]:
df['dob_years'].value_counts().sort_index()

0     101
19     14
20     51
21    111
22    183
23    254
24    264
25    357
26    408
27    493
28    503
29    545
30    540
31    560
32    510
33    581
34    603
35    617
36    555
37    537
38    598
39    573
40    609
41    607
42    597
43    513
44    547
45    497
46    475
47    480
48    538
49    508
50    514
51    448
52    484
53    459
54    479
55    443
56    487
57    460
58    461
59    444
60    377
61    355
62    352
63    269
64    265
65    194
66    183
67    167
68     99
69     85
70     65
71     58
72     33
73      8
74      6
75      1
Name: dob_years, dtype: int64

In [33]:
df['dob_years'].describe()

count    21525.000000
mean        43.293380
std         12.574584
min          0.000000
25%         33.000000
50%         42.000000
75%         53.000000
max         75.000000
Name: dob_years, dtype: float64

In [34]:
df_test = df.loc[df['dob_years'] == 0]
df_test

children  days_employed  dob_years            education  education_id  \
99            0           2194          0  Secondary Education             1   
149           0           2664          0  secondary education             1   
270           3           1872          0  secondary education             1   
578           0           2194          0  secondary education             1   
1040          0           1158          0    bachelor's degree             0   
...         ...            ...        ...                  ...           ...   
19829         0           2194          0  secondary education             1   
20462         0           2194          0  secondary education             1   
20577         0           2194          0  secondary education             1   
21179         2            108          0    bachelor's degree             0   
21313         0           1268          0  secondary education             1   

      family_status  family_status_id gender income_type  debt  total_income  \
99          married                 0      F     retiree     0         11406   
149        divorced                 3      F    employee     0         11228   
270         married                 0      F    employee     0         16346   
578         married                 0      F     retiree     0         15619   
1040       divorced                 3      F    business     0         48639   
...             ...               ...    ...         ...   ...           ...   
19829       married                 0      F    employee     0         23202   
20462       married                 0      F     retiree     0         41471   
20577     unmarried                 4      F     retiree     0         20766   
21179       married                 0      M    business     0         38512   
21313     unmarried                 4      M    employee     0         24387   

                            purpose  
99                              car  
149            housing transactions  
270              housing renovation  
578    construction of own property  
1040                   to own a car  
...                             ...  
19829                       housing  
20462      purchase of my own house  
20577                      property  
21179        building a real estate  
21313          housing transactions  

[101 rows x 12 columns]

In [35]:
print(df['dob_years'].agg('mean'))
print(df['dob_years'].agg('median'))
print(df['dob_years'].agg('std'))

43.29337979094077
42.0
12.574583857924216


**Step 2.3.2**

Menindaklanjuti nilai yang tidak rasional (0) dengan mengganti nilainya.
Nilai yang dipilih untuk menggantikan adalah median.

In [36]:
df.loc[df['dob_years'] == 0 , 'dob_years'] = df['dob_years'].median()

In [37]:
df['dob_years'].value_counts().sort_index()

19     14
20     51
21    111
22    183
23    254
24    264
25    357
26    408
27    493
28    503
29    545
30    540
31    560
32    510
33    581
34    603
35    617
36    555
37    537
38    598
39    573
40    609
41    607
42    698
43    513
44    547
45    497
46    475
47    480
48    538
49    508
50    514
51    448
52    484
53    459
54    479
55    443
56    487
57    460
58    461
59    444
60    377
61    355
62    352
63    269
64    265
65    194
66    183
67    167
68     99
69     85
70     65
71     58
72     33
73      8
74      6
75      1
Name: dob_years, dtype: int64

In [38]:
df['dob_years'].describe()

count    21525.000000
mean        43.490453
std         12.218595
min         19.000000
25%         34.000000
50%         42.000000
75%         53.000000
max         75.000000
Name: dob_years, dtype: float64

### Mengelola kolom 'education' & 'education_id' <a id='kol_education'></a>

**Step 2.4.1**

Mengetahui variasi dan jumlah data dari kolom 'education' & 'education_id' yang diurutkan berdasarkan index.

In [39]:
print(df[['education','education_id']].value_counts())

education            education_id
secondary education  1               13750
bachelor's degree    0                4718
SECONDARY EDUCATION  1                 772
Secondary Education  1                 711
some college         2                 668
BACHELOR'S DEGREE    0                 274
Bachelor's Degree    0                 268
primary education    3                 250
Some College         2                  47
SOME COLLEGE         2                  29
PRIMARY EDUCATION    3                  17
Primary Education    3                  15
graduate degree      4                   4
GRADUATE DEGREE      4                   1
Graduate Degree      4                   1
dtype: int64


**Step 2.4.2**

Merapikan duplikasi data pada kolom 'education' yang disebabkan penulisan *letter case* yang tidak konsisten.

In [40]:
df['education'] = df['education'].str.lower()
print(df[['education','education_id']].value_counts())

education            education_id
secondary education  1               15233
bachelor's degree    0                5260
some college         2                 744
primary education    3                 282
graduate degree      4                   6
dtype: int64


### Mengelola kolom 'family_status_id' & 'family_status' <a id='kol_famstat'></a>

**Step 2.5.1**

Mengetahui variasi dan jumlah data dari kolom 'family_status_id' & 'family_status' yang diurutkan berdasarkan index.

In [41]:
print(df[['family_status_id','family_status']].value_counts().sort_index())

family_status_id  family_status    
0                 married              12380
1                 civil partnership     4177
2                 widow / widower        960
3                 divorced              1195
4                 unmarried             2813
dtype: int64


### Mengelola kolom 'gender' <a id='kol_gender'></a>

**Step 2.6.1**

Mengetahui variasi dan jumlah data dari kolom 'gender' yang diurutkan berdasarkan index.

In [42]:
df['gender'].value_counts().sort_index()

F      14236
M       7288
XNA        1
Name: gender, dtype: int64

In [43]:
df_test = df.loc[df['gender'] == 'XNA']
df_test

children  days_employed  dob_years     education  education_id  \
10701         0           2358         24  some college             2   

           family_status  family_status_id gender income_type  debt  \
10701  civil partnership                 1    XNA    business     0   

       total_income          purpose  
10701         32624  buy real estate

**Step 2.6.2**

Mengubah data yang tidak jelas (XNA) dengan data yang sudah lazim (M/F). Data pengganti yang dipilih adalah berdasarkab modus.

In [44]:
print(df['gender'].mode()[0])

F


In [45]:
df.loc[df['gender'] == 'XNA', 'gender'] = df['gender'].mode()[0]

In [46]:
df['gender'].value_counts()

F    14237
M     7288
Name: gender, dtype: int64

### Mengelola kolom 'income_type' - 'debt' - 'purpose' <a id='kol_inctype_debt_purpose'></a>

**Step 2.7.1**

Mengetahui variasi dan jumlah data dari kolom 'income_type','debt', dan 'purpose' yang diurutkan berdasarkan index.

In [47]:
df['income_type'].value_counts().sort_index()

business                        5085
civil servant                   1459
employee                       11119
entrepreneur                       2
paternity / maternity leave        1
retiree                         3856
student                            1
unemployed                         2
Name: income_type, dtype: int64

In [48]:
df['debt'].value_counts().sort_index()

0    19784
1     1741
Name: debt, dtype: int64

In [49]:
df['purpose'].value_counts().sort_index()

building a property                         620
building a real estate                      626
buy commercial real estate                  664
buy real estate                             624
buy residential real estate                 607
buying a second-hand car                    479
buying my own car                           505
buying property for renting out             653
car                                         495
car purchase                                462
cars                                        478
construction of own property                635
education                                   447
getting an education                        443
getting higher education                    426
going to university                         496
having a wedding                            777
housing                                     647
housing renovation                          612
housing transactions                        653
profile education                       

**Notes:**

Terlalu banyak kategori dalam kolom 'purpose'. Selanjutnya akan dilakukan *grouping* untuk mempermudah analisis.

### Menangani duplikasi data <a id='duplikasi_data'></a>

**Step 2.8.1**

Setelah proses penggantian nilai yang hilang dan proses pembersihan data lainnya telah dilakukan, selanjutnya melakukan penanganan duplikasi data.

In [50]:
print(df.duplicated().sum())

72


In [51]:
df_dup = df.loc[df.duplicated()]
df_dup

children  days_employed  dob_years            education  education_id  \
2849          0           2194         41  secondary education             1   
3290          0           2194         58  secondary education             1   
4182          1           2194         34    bachelor's degree             0   
4851          0           2194         60  secondary education             1   
5557          0           2194         58  secondary education             1   
...         ...            ...        ...                  ...           ...   
20702         0           2194         64  secondary education             1   
21032         0           2194         60  secondary education             1   
21132         0           2194         47  secondary education             1   
21281         1           2194         30    bachelor's degree             0   
21415         0           2194         54  secondary education             1   

           family_status  family_status_id gender income_type  debt  \
2849             married                 0      F    employee     0   
3290   civil partnership                 1      F     retiree     0   
4182   civil partnership                 1      F    employee     0   
4851   civil partnership                 1      F     retiree     0   
5557   civil partnership                 1      F     retiree     0   
...                  ...               ...    ...         ...   ...   
20702            married                 0      F     retiree     0   
21032            married                 0      F     retiree     0   
21132            married                 0      F    employee     0   
21281            married                 0      F    employee     0   
21415            married                 0      F     retiree     0   

       total_income                              purpose  
2849          23202  purchase of the house for my family  
3290          23202                    to have a wedding  
4182          23202                     wedding ceremony  
4851          23202                     wedding ceremony  
5557          23202                    to have a wedding  
...             ...                                  ...  
20702         23202              supplementary education  
21032         23202                   to become educated  
21132         23202                   housing renovation  
21281         23202           buy commercial real estate  
21415         23202                 housing transactions  

[72 rows x 12 columns]

In [52]:
df

children  days_employed  dob_years            education  education_id  \
0             1           8437         42    bachelor's degree             0   
1             1           4024         36  secondary education             1   
2             0           5623         33  secondary education             1   
3             3           4124         32  secondary education             1   
4             0           2194         53  secondary education             1   
...         ...            ...        ...                  ...           ...   
21520         1           4529         43  secondary education             1   
21521         0           2194         67  secondary education             1   
21522         1           2113         38  secondary education             1   
21523         3           3112         38  secondary education             1   
21524         2           1984         40  secondary education             1   

           family_status  family_status_id gender income_type  debt  \
0                married                 0      F    employee     0   
1                married                 0      F    employee     0   
2                married                 0      M    employee     0   
3                married                 0      M    employee     0   
4      civil partnership                 1      F     retiree     0   
...                  ...               ...    ...         ...   ...   
21520  civil partnership                 1      F    business     0   
21521            married                 0      F     retiree     0   
21522  civil partnership                 1      M    employee     1   
21523            married                 0      M    employee     1   
21524            married                 0      F    employee     0   

       total_income                  purpose  
0             40620    purchase of the house  
1             17932             car purchase  
2             23341    purchase of the house  
3             42820  supplementary education  
4             25378        to have a wedding  
...             ...                      ...  
21520         35966     housing transactions  
21521         24959        purchase of a car  
21522         14347                 property  
21523         39054        buying my own car  
21524         13127             to buy a car  

[21525 rows x 12 columns]

**Step 2.8.2**

Menghapus duplikasi data dan melakukan reset index.

In [53]:
df = df.drop_duplicates(ignore_index=True)
df

children  days_employed  dob_years            education  education_id  \
0             1           8437         42    bachelor's degree             0   
1             1           4024         36  secondary education             1   
2             0           5623         33  secondary education             1   
3             3           4124         32  secondary education             1   
4             0           2194         53  secondary education             1   
...         ...            ...        ...                  ...           ...   
21448         1           4529         43  secondary education             1   
21449         0           2194         67  secondary education             1   
21450         1           2113         38  secondary education             1   
21451         3           3112         38  secondary education             1   
21452         2           1984         40  secondary education             1   

           family_status  family_status_id gender income_type  debt  \
0                married                 0      F    employee     0   
1                married                 0      F    employee     0   
2                married                 0      M    employee     0   
3                married                 0      M    employee     0   
4      civil partnership                 1      F     retiree     0   
...                  ...               ...    ...         ...   ...   
21448  civil partnership                 1      F    business     0   
21449            married                 0      F     retiree     0   
21450  civil partnership                 1      M    employee     1   
21451            married                 0      M    employee     1   
21452            married                 0      F    employee     0   

       total_income                  purpose  
0             40620    purchase of the house  
1             17932             car purchase  
2             23341    purchase of the house  
3             42820  supplementary education  
4             25378        to have a wedding  
...             ...                      ...  
21448         35966     housing transactions  
21449         24959        purchase of a car  
21450         14347                 property  
21451         39054        buying my own car  
21452         13127             to buy a car  

[21453 rows x 12 columns]

### Mengategorikan data <a id='kategori_data'></a>
Setelah proses pembersihan data dilakukan, selanjutnya melakukan kategorisasi data yang diperlukan untuk proses analisis.

**Step 2.9.1**

Melakukan kategorisasi data **kolom 'children'**

In [54]:
def assign_have_kids(kids):
    if kids < 0 or pd.isna(kids):
        return 'NA'
    elif kids == 0:
        return 'No'
    else:
        return 'Yes'

df['have_kids'] = df['children'].apply(assign_have_kids)

C:\Users\Luqmanulhakim\AppData\Local\Temp\ipykernel_26948\2296291625.py:9: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df['have_kids'] = df['children'].apply(assign_have_kids)


In [55]:
df

children  days_employed  dob_years            education  education_id  \
0             1           8437         42    bachelor's degree             0   
1             1           4024         36  secondary education             1   
2             0           5623         33  secondary education             1   
3             3           4124         32  secondary education             1   
4             0           2194         53  secondary education             1   
...         ...            ...        ...                  ...           ...   
21448         1           4529         43  secondary education             1   
21449         0           2194         67  secondary education             1   
21450         1           2113         38  secondary education             1   
21451         3           3112         38  secondary education             1   
21452         2           1984         40  secondary education             1   

           family_status  family_status_id gender income_type  debt  \
0                married                 0      F    employee     0   
1                married                 0      F    employee     0   
2                married                 0      M    employee     0   
3                married                 0      M    employee     0   
4      civil partnership                 1      F     retiree     0   
...                  ...               ...    ...         ...   ...   
21448  civil partnership                 1      F    business     0   
21449            married                 0      F     retiree     0   
21450  civil partnership                 1      M    employee     1   
21451            married                 0      M    employee     1   
21452            married                 0      F    employee     0   

       total_income                  purpose have_kids  
0             40620    purchase of the house       Yes  
1             17932             car purchase       Yes  
2             23341    purchase of the house        No  
3             42820  supplementary education       Yes  
4             25378        to have a wedding        No  
...             ...                      ...       ...  
21448         35966     housing transactions       Yes  
21449         24959        purchase of a car        No  
21450         14347                 property       Yes  
21451         39054        buying my own car       Yes  
21452         13127             to buy a car       Yes  

[21453 rows x 13 columns]

**Step 2.9.2**

Melakukan kategorisasi data **kolom 'dob_years'**

In [56]:
def assign_age_group(age):
    if age < 0 or pd.isna(age):
        return 'NA'
    elif age < 10:
        return '0-9'
    elif age < 20:
        return '10-19'
    elif age < 30:
        return '20-29'
    elif age < 40:
        return '30-39'
    elif age < 50:
        return '40-49'
    elif age < 60:
        return '50-59'
    elif age < 70:
        return '60-69'
    else:
        return '70+'

df['age_group'] = df['dob_years'].apply(assign_age_group)

C:\Users\Luqmanulhakim\AppData\Local\Temp\ipykernel_26948\3215841067.py:21: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df['age_group'] = df['dob_years'].apply(assign_age_group)


In [57]:
df

children  days_employed  dob_years            education  education_id  \
0             1           8437         42    bachelor's degree             0   
1             1           4024         36  secondary education             1   
2             0           5623         33  secondary education             1   
3             3           4124         32  secondary education             1   
4             0           2194         53  secondary education             1   
...         ...            ...        ...                  ...           ...   
21448         1           4529         43  secondary education             1   
21449         0           2194         67  secondary education             1   
21450         1           2113         38  secondary education             1   
21451         3           3112         38  secondary education             1   
21452         2           1984         40  secondary education             1   

           family_status  family_status_id gender income_type  debt  \
0                married                 0      F    employee     0   
1                married                 0      F    employee     0   
2                married                 0      M    employee     0   
3                married                 0      M    employee     0   
4      civil partnership                 1      F     retiree     0   
...                  ...               ...    ...         ...   ...   
21448  civil partnership                 1      F    business     0   
21449            married                 0      F     retiree     0   
21450  civil partnership                 1      M    employee     1   
21451            married                 0      M    employee     1   
21452            married                 0      F    employee     0   

       total_income                  purpose have_kids age_group  
0             40620    purchase of the house       Yes     40-49  
1             17932             car purchase       Yes     30-39  
2             23341    purchase of the house        No     30-39  
3             42820  supplementary education       Yes     30-39  
4             25378        to have a wedding        No     50-59  
...             ...                      ...       ...       ...  
21448         35966     housing transactions       Yes     40-49  
21449         24959        purchase of a car        No     60-69  
21450         14347                 property       Yes     30-39  
21451         39054        buying my own car       Yes     30-39  
21452         13127             to buy a car       Yes     40-49  

[21453 rows x 14 columns]

**Step 2.9.3**

Melakukan kategorisasi data **kolom 'purpose'**

In [58]:
wedding = ['wedding ceremony','having a wedding','to have a wedding']
property = ['real estate transactions','buy commercial real estate','buying property for renting out','housing transactions',
            'transactions with commercial real estate','purchase of the house','housing','purchase of the house for my family',
            'construction of own property','property','transactions with my real estate','building a real estate',
            'buy real estate','building a property','purchase of my own house','housing renovation','buy residential real estate']
vehicle = ['buying my own car','car','second-hand car purchase','to own a car','buying a second-hand car','cars',
          'to buy a car','car purchase','purchase of a car']
education = ['going to university','supplementary education','university education','education','to get a supplementary education',
            'getting an education','profile education','getting higher education','to become educated']


def assign_purpose_class(purpose):
    if purpose in wedding:
        return 'wedding'
    elif purpose in property:
        return 'property'
    elif purpose in vehicle:
        return 'vehicle'
    elif purpose in education:
        return 'education'
    else:
        return 'other'

df['used_for'] = df['purpose'].apply(assign_purpose_class)


C:\Users\Luqmanulhakim\AppData\Local\Temp\ipykernel_26948\80679556.py:24: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df['used_for'] = df['purpose'].apply(assign_purpose_class)


In [59]:
df

children  days_employed  dob_years            education  education_id  \
0             1           8437         42    bachelor's degree             0   
1             1           4024         36  secondary education             1   
2             0           5623         33  secondary education             1   
3             3           4124         32  secondary education             1   
4             0           2194         53  secondary education             1   
...         ...            ...        ...                  ...           ...   
21448         1           4529         43  secondary education             1   
21449         0           2194         67  secondary education             1   
21450         1           2113         38  secondary education             1   
21451         3           3112         38  secondary education             1   
21452         2           1984         40  secondary education             1   

           family_status  family_status_id gender income_type  debt  \
0                married                 0      F    employee     0   
1                married                 0      F    employee     0   
2                married                 0      M    employee     0   
3                married                 0      M    employee     0   
4      civil partnership                 1      F     retiree     0   
...                  ...               ...    ...         ...   ...   
21448  civil partnership                 1      F    business     0   
21449            married                 0      F     retiree     0   
21450  civil partnership                 1      M    employee     1   
21451            married                 0      M    employee     1   
21452            married                 0      F    employee     0   

       total_income                  purpose have_kids age_group   used_for  
0             40620    purchase of the house       Yes     40-49   property  
1             17932             car purchase       Yes     30-39    vehicle  
2             23341    purchase of the house        No     30-39   property  
3             42820  supplementary education       Yes     30-39  education  
4             25378        to have a wedding        No     50-59    wedding  
...             ...                      ...       ...       ...        ...  
21448         35966     housing transactions       Yes     40-49   property  
21449         24959        purchase of a car        No     60-69    vehicle  
21450         14347                 property       Yes     30-39   property  
21451         39054        buying my own car       Yes     30-39    vehicle  
21452         13127             to buy a car       Yes     40-49    vehicle  

[21453 rows x 15 columns]

**Step 2.9.4**

Melakukan kategorisasi data **kolom 'total_income'**

In [60]:
df['total_income'].describe()

count     21453.000000
mean      26435.800121
std       15683.720829
min        3306.000000
25%       17219.000000
50%       23202.000000
75%       31331.000000
max      362496.000000
Name: total_income, dtype: float64

In [61]:
print(df['total_income'].loc[df['total_income'] < 10000].sort_values())
print('')
print(df['total_income'].loc[(df['total_income'] >= 10000) & (df['total_income'] < 20000)].sort_values())
print('')
print(df['total_income'].loc[(df['total_income'] >= 20000) & (df['total_income'] < 30000)].sort_values())
print('')
print(df['total_income'].loc[(df['total_income'] >= 30000) & (df['total_income'] < 40000)].sort_values())
print('')
print(df['total_income'].loc[(df['total_income'] >= 40000) & (df['total_income'] < 50000)].sort_values())
print('')
print(df['total_income'].loc[df['total_income'] >= 50000].sort_values())

14555    3306
12983    3392
16137    3418
1598     3471
14247    3503
         ... 
14301    9992
9567     9992
647      9993
783      9995
10740    9996
Name: total_income, Length: 926, dtype: int32

3373     10000
1742     10001
12495    10004
7581     10006
2787     10011
         ...  
7046     19994
12725    19997
2125     19998
17401    19998
1857     19999
Name: total_income, Length: 6443, dtype: int32

10764    20000
15695    20001
12484    20003
21373    20003
7661     20006
         ...  
15963    29989
3914     29991
10477    29991
7818     29992
10982    29992
Name: total_income, Length: 8165, dtype: int32

14861    30009
14338    30010
20195    30011
11894    30013
16559    30013
         ...  
19725    39987
19212    39990
6760     39995
4839     39995
20699    39998
Name: total_income, Length: 3107, dtype: int32

17159    40020
4031     40036
20671    40047
3098     40050
13393    40050
         ...  
4355     49934
3875     49963
6415     49965
13537    49981
10089    4

In [62]:
def assign_inc_group(income):
    if income < 0 or pd.isna(income):
        return 'NA'
    elif income < 10000:
        return '0-9999'
    elif income < 20000:
        return '10000-19999'
    elif income < 30000:
        return '20000-29999'
    elif income < 40000:
        return '30000-39999'
    elif income < 50000:
        return '40000-49999'
    else:
        return '50000 and up'

df['inc_group'] = df['total_income'].apply(assign_inc_group)

C:\Users\Luqmanulhakim\AppData\Local\Temp\ipykernel_26948\3534178982.py:17: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df['inc_group'] = df['total_income'].apply(assign_inc_group)


In [63]:
df

children  days_employed  dob_years            education  education_id  \
0             1           8437         42    bachelor's degree             0   
1             1           4024         36  secondary education             1   
2             0           5623         33  secondary education             1   
3             3           4124         32  secondary education             1   
4             0           2194         53  secondary education             1   
...         ...            ...        ...                  ...           ...   
21448         1           4529         43  secondary education             1   
21449         0           2194         67  secondary education             1   
21450         1           2113         38  secondary education             1   
21451         3           3112         38  secondary education             1   
21452         2           1984         40  secondary education             1   

           family_status  family_status_id gender income_type  debt  \
0                married                 0      F    employee     0   
1                married                 0      F    employee     0   
2                married                 0      M    employee     0   
3                married                 0      M    employee     0   
4      civil partnership                 1      F     retiree     0   
...                  ...               ...    ...         ...   ...   
21448  civil partnership                 1      F    business     0   
21449            married                 0      F     retiree     0   
21450  civil partnership                 1      M    employee     1   
21451            married                 0      M    employee     1   
21452            married                 0      F    employee     0   

       total_income                  purpose have_kids age_group   used_for  \
0             40620    purchase of the house       Yes     40-49   property   
1             17932             car purchase       Yes     30-39    vehicle   
2             23341    purchase of the house        No     30-39   property   
3             42820  supplementary education       Yes     30-39  education   
4             25378        to have a wedding        No     50-59    wedding   
...             ...                      ...       ...       ...        ...   
21448         35966     housing transactions       Yes     40-49   property   
21449         24959        purchase of a car        No     60-69    vehicle   
21450         14347                 property       Yes     30-39   property   
21451         39054        buying my own car       Yes     30-39    vehicle   
21452         13127             to buy a car       Yes     40-49    vehicle   

         inc_group  
0      40000-49999  
1      10000-19999  
2      20000-29999  
3      40000-49999  
4      20000-29999  
...            ...  
21448  30000-39999  
21449  20000-29999  
21450  10000-19999  
21451  30000-39999  
21452  10000-19999  

[21453 rows x 16 columns]

### Kesimpulan <a id='data_preprocessing_conclusions'></a>

* Data yang hilang mencapai 10% dari 2 kolom secara bersamaan yakni 2174 baris dari 21525 baris pada kolom 'days_employed' dan 'total income'. Selain itu pada kolom 'days_employed' terdapat 3445 baris atau 16% dari total baris dengan nilai yang tidak rasional, di samping juga terdapat nilai minus yang perlu diabsolutkan.
* Data minus juga terdapat pada kolom 'children', di samping juga terdapat data irrasional dengan nilai 20.
* Penginputan dengan *letter case* yang tidak seragam masih terdapat pada kolom 'education' yang masih relatif mudah untuk disesuaikan dan tidak mengubah esensi data.
* Meskipun sedikit (0,5%), sebanyak 101 data dalam kolom 'dob_years' masih menampilkan nilai 0, serta untuk kolom 'gender' hanya terdapat 1 data saja yang berada di luar Male/Female, sehingga relatif mudah untuk disesuaikan dan tidak terlalu berdampak signifikan secara keseluruhan.
* Duplikasi data masih ada namun tidak terlalu signifikan, yakni sebesar 0,33% atau 72 baris saja.

[Kembali ke Daftar Isi](#back)

## Tahap 3. Analisis Hubungan <a id='analisis_hubungan'></a>
Melakukan analisis beberapa faktor terhadap probabilitas seseorang melakukan gagal bayar pinjaman.

### Hubungan 1: memiliki anak dan probabilitas gagal bayar <a id='analisis_hub_1'></a>
**Step 3.1**

Menganalisis hubungan antara *memiliki anak* dan *probabilitas seseorang melakukan gagal bayar pinjaman*.

In [64]:
print(df.pivot_table(
    index='have_kids',
    columns='gender',
    values='debt',
    aggfunc='count',
    margins=True)
)
print('')
print(df.pivot_table(
    index='have_kids',
    columns='gender',
    values='debt',
    aggfunc='sum',
    margins=True)
)

gender         F     M    All
have_kids                    
No          9520  4570  14090
Yes         4654  2709   7363
All        14174  7279  21453

gender       F    M   All
have_kids                
No         592  471  1063
Yes        402  276   678
All        994  747  1741


In [65]:
df_kids_all = df.pivot_table(
    index='have_kids',
    columns='gender',
    values='debt',
    aggfunc='count',
    margins=True)

df_kids_debt = df.pivot_table(
    index='have_kids',
    columns='gender',
    values='debt',
    aggfunc='sum',
    margins=True)

df_kids_pctg = round((df_kids_debt / df_kids_all)*100,2)
df_kids_pctg

gender        F      M   All
have_kids                   
No         6.22  10.31  7.54
Yes        8.64  10.19  9.21
All        7.01  10.26  8.12

**Hasil Analisis:**

Dari analisis di atas dapat diambil kesimpulan sebagai berikut:
* Nasabah yang memiliki anak memiliki potensi gagal bayar pinjaman lebih besar dibandingkan nasabah yang tidak/belum memiliki anak, dengan persentase 9,21% berbanding 7,54%. 
* Jika dilakukan drilldown dari sisi gender, nasabah wanita yang memiliki anak juga memiliki potensi gagal bayar pinjaman lebih besar dibandingkan nasabah wanita yang tidak/belum memiliki anak, dengan persentase 8,64% berbanding 6,22%. 
* Berbeda dengan nasabah pria yang justru memiliki potensi gagal bayar pinjaman lebih besar ada pada nasabah pria yang tidak/belum memiliki anak, dengan persentase 10,31% berbanding 10,19%.

### Hubungan 2: status perkawinan dan probabilitas gagal bayar <a id='analisis_hub_2'></a>
**Step 3.2**

Menganalisis hubungan antara *status perkawinan* dan *probabilitas seseorang melakukan gagal bayar pinjaman*.

In [66]:
print(df.pivot_table(
    index='family_status',
    columns='gender',
    values='debt',
    aggfunc='count',
    margins=True)
)
print('')
print(df.pivot_table(
    index='family_status',
    columns='gender',
    values='debt',
    aggfunc='sum',
    margins=True)
)

gender                 F     M    All
family_status                        
civil partnership   2845  1305   4150
divorced             936   259   1195
married             7760  4579  12339
unmarried           1729  1081   2810
widow / widower      904    55    959
All                14174  7279  21453

gender               F    M   All
family_status                    
civil partnership  233  155   388
divorced            61   24    85
married            530  401   931
unmarried          118  156   274
widow / widower     52   11    63
All                994  747  1741


In [67]:
df_fam_all = df.pivot_table(
    index='family_status',
    columns='gender',
    values='debt',
    aggfunc='count',
    margins=True)

df_fam_debt = df.pivot_table(
    index='family_status',
    columns='gender',
    values='debt',
    aggfunc='sum',
    margins=True)

df_fam_pctg = round((df_fam_debt / df_fam_all)*100,2)
df_fam_pctg

gender                F      M   All
family_status                       
civil partnership  8.19  11.88  9.35
divorced           6.52   9.27  7.11
married            6.83   8.76  7.55
unmarried          6.82  14.43  9.75
widow / widower    5.75  20.00  6.57
All                7.01  10.26  8.12

**Hasil Analisis:**

Dari analisis di atas dapat diambil kesimpulan sebagai berikut:
* Nasabah yang tidak/belum menikah, baik berstatus unmarried maupun civil partnership memiliki potensi gagal bayar pinjaman lebih besar dibandingkan nasabah yang telah menikah, baik berstatus married, divorced, maupun widow/widower. Hal ini ditunjukkan dengan rasio 9,75% dan 9,35% (keduanya di atas rasio debt dari outstanding keseluruhan yakni 8,12%) dibandingkan dengan 7,55%, 7,11%, dan 6,57%.

### Hubungan 3: tingkat pendapatan dan probabilitas gagal bayar <a id='analisis_hub_3'></a>
**Step 3.3**

Menganalisis hubungan antara *tingkat pendapatan* dan *probabilitas seseorang melakukan gagal bayar pinjaman*.

In [68]:
print(df.pivot_table(
    index='inc_group',
    columns='gender',
    values='debt',
    aggfunc='count',
    margins=True)
)
print('')
print(df.pivot_table(
    index='inc_group',
    columns='gender',
    values='debt',
    aggfunc='sum',
    margins=True)
)

gender            F     M    All
inc_group                       
0-9999          794   132    926
10000-19999    4852  1591   6443
20000-29999    5319  2846   8165
30000-39999    1745  1362   3107
40000-49999     808   684   1492
50000 and up    656   664   1320
All           14174  7279  21453

gender          F    M   All
inc_group                   
0-9999         45   13    58
10000-19999   363  187   550
20000-29999   388  309   697
30000-39999   104  138   242
40000-49999    52   50   102
50000 and up   42   50    92
All           994  747  1741


In [69]:
df_inc_all = df.pivot_table(
    index='inc_group',
    columns='gender',
    values='debt',
    aggfunc='count',
    margins=True)

df_inc_debt = df.pivot_table(
    index='inc_group',
    columns='gender',
    values='debt',
    aggfunc='sum',
    margins=True)

df_inc_pctg = round((df_inc_debt / df_inc_all)*100,2)
df_inc_pctg

gender           F      M   All
inc_group                      
0-9999        5.67   9.85  6.26
10000-19999   7.48  11.75  8.54
20000-29999   7.29  10.86  8.54
30000-39999   5.96  10.13  7.79
40000-49999   6.44   7.31  6.84
50000 and up  6.40   7.53  6.97
All           7.01  10.26  8.12

**Hasil Analisis:**

Dari analisis di atas dapat diambil kesimpulan sebagai berikut:
* Nasabah dengan rentang income 10.000 - 29999 memiliki potensi gagal bayar pinjaman lebih besar dibandingkan nasabah dengan rentang income lainnya, baik yang lebih kecil (0 - 9.999) maupun yang lebih besar, yakni dengan rasio 8,54% (di atas rasio debt dari outstanding keseluruhan yakni 8,12%).
* Dari segi gender juga menampilkan data yang sejalan, di mana baik nasabah pria maupun nasabah wanita dengan rentang income 10.000 - 29999 memiliki persentase yang lebih besar dibandingkan rentang income lainnya.

### Hubungan 4: tujuan pinjaman dan probabilitas gagal bayar <a id='analisis_hub_4'></a>
**Step 4.4**

Menganalisis hubungan antara *tujuan pinjaman* dan *probabilitas seseorang melakukan gagal bayar pinjaman*.

In [70]:
print(df.pivot_table(
    index='used_for',
    columns='gender',
    values='debt',
    aggfunc='count',
    margins=True)
)
print('')
print(df.pivot_table(
    index='used_for',
    columns='gender',
    values='debt',
    aggfunc='sum',
    margins=True)
)

gender         F     M    All
used_for                     
education   2667  1346   4013
property    7086  3725  10811
vehicle     2860  1446   4306
wedding     1561   762   2323
All        14174  7279  21453

gender       F    M   All
used_for                 
education  207  163   370
property   444  338   782
vehicle    235  168   403
wedding    108   78   186
All        994  747  1741


In [71]:
df_used_all = df.pivot_table(
    index='used_for',
    columns='gender',
    values='debt',
    aggfunc='count',
    margins=True)

df_used_debt = df.pivot_table(
    index='used_for',
    columns='gender',
    values='debt',
    aggfunc='sum',
    margins=True)

df_used_pctg = round((df_used_debt / df_used_all)*100,2)
df_used_pctg

gender        F      M   All
used_for                    
education  7.76  12.11  9.22
property   6.27   9.07  7.23
vehicle    8.22  11.62  9.36
wedding    6.92  10.24  8.01
All        7.01  10.26  8.12

**Hasil Analisis:**

Dari analisis di atas dapat diambil kesimpulan sebagai berikut:
* Tujuan pinjaman untuk keperluan kendaraan (vehicle) dan pendidikan (education) memiliki potensi gagal bayar pinjaman lebih besar dibandingkan untuk keperluan perumahan (property) dan pernikahan (wedding). Hal ini ditunjukkan dengan rasio 9,36% dan 9,22% (keduanya di atas rasio debt dari outstanding keseluruhan yakni 8,12%) dibandingkan dengan 7,23% dan 8,01%. 
* Dari segi gender juga menampilkan data yang sejalan, di mana baik nasabah pria maupun nasabah wanita dengan tujuan pinjaman untuk keperluan kendaraan (vehicle) dan pendidikan (education) juga memiliki potensi gagal bayar pinjaman lebih besar dibandingkan untuk keperluan perumahan (property) dan pernikahan (wedding). Untuk pria dengan rasio 11,62% dan 12,11% dibandingkan dengan 9,07% dan 10,24%. Untuk wanita dengan rasio 8,22% dan 7,76% dibandingkan dengan 6,27% dan 6,92%. 

## Temuan <a id='end'></a>

* Faktor-faktor yang mempengaruhi peningkatan potensi gagal bayar pinjaman yakni nasabah yang memiliki anak, nasabah yang tidak/belum menikah, dengan income di kisaran 10000-29999, dan tujuan pinjamannya untuk keperluan kendaraan dan pendidikan.
* Dari faktor-faktor di atas, terdapat dua faktor yang secara natural bisa menjadi pertanyaan lanjutan, yakni nasabah yang memiliki anak dan nasabah yang tidak/belum menikah. Hal ini terlihat seperti kontradiktif namun tetap ada ruang di tengahnya yakni nasabah yang telah menikah namun belum memiliki anak menjadi faktor yang tidak mempengaruhi peningkatan potensi gagal bayar pinjaman.
* Terlepas dari simpulan di atas, hal lain yang juga amat penting adalah banyaknya data yang memerlukan *adjustment* baik karena hilang maupun tidak rasional. Duplikasi data masih ada namun tidak terlalu signifikan, yakni sebesar 0,33% atau 72 baris saja.
* Data yang hilang mencapai 10% dari 2 kolom secara bersamaan yakni 2174 baris dari 21525 baris pada kolom 'days_employed' dan 'total income'. Selain itu pada kolom 'days_employed' terdapat 3445 baris atau 16% dari total baris dengan nilai yang tidak rasional, di samping juga terdapat nilai minus yang perlu diabsolutkan.
* Data minus juga terdapat pada kolom 'children', di samping juga terdapat data irrasional dengan nilai 20.
* Penginputan dengan *letter case* yang tidak seragam masih terdapat pada kolom 'education' yang masih relatif mudah untuk disesuaikan dan tidak mengubah esensi data.
* Meskipun sedikit (0,5%), sebanyak 101 data dalam kolom 'dob_years' masih menampilkan nilai 0, serta untuk kolom 'gender' hanya terdapat 1 data saja yang berada di luar Male/Female, sehingga relatif mudah untuk disesuaikan dan tidak terlalu berdampak signifikan secara keseluruhan.
* Mengingat data yang memerlukan perbaikan pada saat proses analisis mencakup proporsi data yang cukup besar, fakta ini bisa menjadi dasar untuk mempertanyakan reliabilitas kesimpulan yang dihasilkan.

[Kembali ke Daftar Isi](#back)